### ISSUE TITLE
Provide means of safely updating all bundles

### ISSUE THREAD

**Issue by Lange** [nodecg/nodecg#162]
I maintain a lot of production NodeCG instances. I often find myself wanting to update all the bundles present on a given instance. The problem is that I have no way of automatically updating all of these bundles to whatever the latest version is that works with that instance's version of NodeCG.

I propose a `cfg/bundles.json` file. This file would be formatted similar to a `package.json`'s `dependencies` field:

``` js
{
  // Gets lfg-filter 0.2.x from GitHub
  "lfg-filter": "https://github.com/SupportClass/lfg-filter.git#~0.2.0",

  // Gets lfg-nucleus 1.x.y from GitHub
  "lfg-nucleus": "https://github.com/SupportClass/lfg-filter.git#^1.1.2",

  // Gets the latest commit of client-bundle from BitBucket
  "client-bundle": "git@bitbucket.org:privateorg/client-bundle.git",

  // Gets npm-bundle@0.6.2 from npm
  "npm-bundle": "0.6.2"
}
```

[`nodecg-cli`](https://github.com/nodecg/nodecg-cli)'s `nodecg install` command would be updated to parse this file and install the appropriate versions of each bundle.
### What would happen when running `nodecg install`?
1. For every bundle:
   1. Get list of releases for the bundle
   2. Figure out if it needs updating (based on the specified semver ranges), if we're installing it from scratch, or if no action is needed
   3. Hold onto the tarball URL for later
2. If any bundles need updating, prompt the user to overwrite
   1. If we're just installing new bundles, no need to prompt for overwrite.
3. Download the tarballs from the URLs we saved earlier
4. Extract all the tarballs to a temp folder
5. Delete the existing bundles
6. Move the extracted bundles to the bundles folder
### Known issues with this approach

This doesn't actually get the latest versions of the bundles that are known to work with the current version of NodeCG. It just gets the latest version that satisfies the supplied semver range. However, this is still a big improvement and the maintainer of the NodeCG instance can do a little legwork to keep those semver ranges where they want them to be.

Ultimately, there are two separate problems here. One is that a NodeCG instance maintainer won't want to download a version of a bundle that breaks their other bundles due to an API change (for example, if lfg-filter updated to 1.0.0 and lfg-nucleus wasn't updated to handle its API changes). The other issue is that the maintainer won't want to download a version of a bundle that doesn't work with their NodeCG version. This proposed approach solves the first problem, and kinda solves the second problem on accident, albeit not in the most direct manner.

Thoughts and feedback are welcome.

**Comment by Lange** [comment]
Upon further thought, this plan of using the GitHub/BitBucket/${otherGitService} APIs to enumerate the list of releases available for a bundle might not be the best idea. I don't think there's an easy-to-use way for me to have the `nodecg install` command get the authentication it needs to read info about private repos.

Both GitHub and BitBucket support HTTPS Basic Auth, which would be kinda easy to set up for the end user, but would require them to keep those credentials in plaintext in a config file somewhere. Not acceptable.

GitHub and BitBucket both also support OAuth, but asking users to create OAuth applications just to use this feature, and asking them to then maintain a config holding their OAuth tokens, seems too cumbersome.

`nodecg install` currently uses `git` for all of its operations. This has the benefit of being able to take advantage of ssh key auth, and it'll leverage any running ssh agents on the system. This feels like the easiest way to authenticate with GH/BB and gain access to a user's private repos. Maybe continuing to use `git` for all these bundle installation and update operations would be best.

tl;dr - Getting release info from GitHub/BitBucket APIs would be ideal, but getting authorization to read private repos is hard. Maybe it's better to just use `git` and ssh agents?

**Comment by Lange** [comment]
I don't think any of the ideas I have for implementing this behavior are good. Ideas welcome.

**Comment by MattMcNam** [comment]
I don't think this should be handled by NodeCG nor the CLI; it should be the responsibility of the instance maintainer to know that the bundles they want to run work with the version of NodeCG they want to run, and there are tons of very well supported, highly configurable deployment tools out there to make sure those right versions get deployed to the server.

I'm working on making my dev (some VMs) & production server setup and deployment automated via Ansible, and for NodeCG bundles I want to try the following methods:
- .tar.gz downloads from a manifest file similar to the format you posted above, either from github, npm or a private sinopia instance
- (private) git repository containing only git sub-repositories pointing to each bundle, cloned into `bundles/` (though this doesn't work for bundles requiring a build step without more work)
- one large archive file containing bundles, extracted to `bundles/` (ugly but guaranteed to be what I expect)

**Comment by Lange** [comment]
Awesome, I'd love to hear what your workflow ends up being. My devops situation right now is a fucking garbage fire so I will absolutely listen to advice.

**Comment by Lange** [comment]
Revisiting this: I agree that if we can use existing third-party tooling for this, then we should. However, we should create a recipe that people can follow to get the desired outcome with a recommended set of tools.

(On that topic, we should add a `recipes` documentation folder in general, [similar to what `ava` has](https://github.com/avajs/ava/tree/master/docs/recipes).)

**Comment by Lange** [comment]
Stale, closing.
